File: libsarc/src/archive.cpp
//
// Copyright (c) 2016 wkmghn.
// 
// Use, modification and distribution is subject to the Boost Software License,
// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
//
#include <sarc/archive.hpp>
#include <cassert>
#include <cstring>

namespace sarc
{

namespace
{
// MagicNumber + Version + NumFiles
const std::uint32_t MINIMUM_ARCHIVE_SIZE = 4 + 4 + 4;

std::uint32_t read_uint32(const uint8_t* data)
{
	return static_cast<std::uint32_t>(data[0]) << 24
		| static_cast<std::uint32_t>(data[1]) << 16
		| static_cast<std::uint32_t>(data[2]) << 8
		| static_cast<std::uint32_t>(data[3]) << 0;
}
}

Archive::Archive(const void* data, std::uint32_t data_size)
	: parsing_result_(ParsingResult::Succeeded)
	, data_(reinterpret_cast<const std::uint8_t*>(data))
	, data_size_(data_size)
{
	if (data_ == nullptr) {
		parsing_result_ = ParsingResult::NullData;
		return;
	}

	if (data_size_ < MINIMUM_ARCHIVE_SIZE) {
		parsing_result_ = ParsingResult::TooFewDataSize;
		return;
	}

	// MagicNumber = 's' 'a' 'r' 'c'
	if (data_[0] != 0x73 || data_[1] != 0x61 || data_[2] != 0x72 || data_[3] != 0x63) {
		parsing_result_ = ParsingResult::DataCorrupted;
		return;
	}

	// Version
	const std::uint32_t version = read_uint32(data_ + 4);
	if (version != 1) {
		parsing_result_ = ParsingResult::UnsupportedVersion;
		return;
	}
}

Archive::~Archive() = default;

std::uint32_t Archive::num_files() const
{
	if (parsing_result_ != ParsingResult::Succeeded) {
		return 0;
	}
	assert(data_ != nullptr);

	return read_uint32(data_ + 8);
}

FileAccessor Archive::get_file(std::uint32_t file_index) const
{
	if (parsing_result_ != ParsingResult::Succeeded) {
		return nullptr;
	}
	assert(data_ != nullptr);

	if (num_files() <= file_index) {
		return nullptr;
	}

	const std::uint32_t offset_offset = 12 + file_index * 4;
	const std::uint32_t file_head_offset = read_uint32(data_ + offset_offset);

	const std::uint8_t* file_head = data_ + file_head_offset;
	// ファイルサイズがゼロでも何らかの値が読み取れる。
	// 少なくともアーカイブ内のどこかを指すが、まともなデータは読み取れない。
	const std::uint32_t file_body_offset_from_head = read_uint32(file_head);
	const std::uint32_t file_body_offset = file_head_offset + file_body_offset_from_head;
	const std::uint8_t* file_body = data_ + file_body_offset;

	const std::uint32_t file_size = read_uint32(file_head + 4);
	// 今のところランタイム側では参照しない
	//std::uint32_t alignment = read_uint32(file_head + 8);
	const char* file_name = reinterpret_cast<const char*>(file_head + 12);

	return FileAccessor(file_body, file_name, file_size);
}

FileAccessor Archive::find_file(const char* file_name) const
{
	if (parsing_result_ != ParsingResult::Succeeded) {
		return nullptr;
	}
	assert(data_ != nullptr);

	const std::uint32_t num = num_files();
	for (std::uint32_t i = 0; i < num; ++i) {
		FileAccessor file = get_file(i);
		assert(file.is_valid());
		if (std::strcmp(file.file_name(), file_name) == 0) {
			return file;
		}
	}

	return nullptr;
}

Archive::Iterator Archive::begin() const
{
	return Iterator(this, 0);
}

Archive::Iterator Archive::end() const
{
	return Iterator(this, num_files());
}

//==============================================================================
// Iterator
//==============================================================================

Archive::Iterator::Iterator()
	: archive_(nullptr)
	, file_index_(0)
{
}

Archive::Iterator::Iterator(const Archive* archive, std::uint32_t file_index)
	: archive_(archive)
	, file_index_(file_index)
{
	assert(archive != nullptr);
	assert(file_index <= archive->num_files());
}

Archive::Iterator::value_type Archive::Iterator::operator*() const
{
	assert(archive_ != nullptr);
	assert(file_index_ < archive_->num_files());
	return archive_->get_file(file_index_);
}

Archive::Iterator::value_type Archive::Iterator::operator[](difference_type index) const
{
	assert(archive_ != nullptr);
	const uint32_t new_file_index = static_cast<std::uint32_t>(file_index_ + index);
	assert(new_file_index <= archive_->num_files());
	return archive_->get_file(new_file_index);
}

Archive::Iterator& Archive::Iterator::operator++()
{
	assert(archive_ != nullptr);
	assert(file_index_ < archive_->num_files());
	++file_index_;
	return *this;
}

Archive::Iterator Archive::Iterator::operator++(int)
{
	assert(archive_ != nullptr);
	assert(file_index_ < archive_->num_files());
	++file_index_;
	return Iterator(archive_, file_index_ - 1);
}

Archive::Iterator& Archive::Iterator::operator--()
{
	assert(archive_ != nullptr);
	assert(0 < file_index_);
	--file_index_;
	return *this;
}

Archive::Iterator Archive::Iterator::operator--(int)
{
	assert(archive_ != nullptr);
	assert(0 < file_index_);
	--file_index_;
	return Iterator(archive_, file_index_ + 1);
}

Archive::Iterator& Archive::Iterator::operator+=(difference_type rhs)
{
	assert(archive_ != nullptr);
	file_index_ = static_cast<std::uint32_t>(file_index_ + rhs);
	assert(file_index_ <= archive_->num_files());
	return *this;
}

Archive::Iterator& Archive::Iterator::operator-=(difference_type rhs)
{
	assert(archive_ != nullptr);
	file_index_ = static_cast<std::uint32_t>(file_index_ - rhs);
	assert(file_index_ <= archive_->num_files());
	return *this;
}

Archive::Iterator operator+(const Archive::Iterator& lhs, Archive::Iterator::difference_type rhs)
{
	assert(lhs.archive_ != nullptr);
	const std::uint32_t file_index = static_cast<std::uint32_t>(lhs.file_index_ + rhs);
	assert(file_index <= lhs.archive_->num_files());
	return Archive::Iterator(lhs.archive_, file_index);
}

Archive::Iterator operator+(Archive::Iterator::difference_type lhs, const Archive::Iterator& rhs)
{
	return rhs + lhs;
}

Archive::Iterator operator-(const Archive::Iterator& lhs, Archive::Iterator::difference_type rhs)
{
	return lhs + (-rhs);
}

Archive::Iterator::difference_type operator-(const Archive::Iterator& lhs, const Archive::Iterator& rhs)
{
	assert(lhs.archive_ != nullptr);
	assert(rhs.archive_ != nullptr);
	return static_cast<Archive::Iterator::difference_type>(lhs.file_index_) - static_cast<Archive::Iterator::difference_type>(rhs.file_index_);
}

bool operator==(const Archive::Iterator& lhs, const Archive::Iterator& rhs)
{
	assert(lhs.archive_ != nullptr);
	assert(rhs.archive_ != nullptr);
	assert(lhs.archive_ == rhs.archive_);
	return lhs.file_index_ == rhs.file_index_;
}

bool operator!=(const Archive::Iterator& lhs, const Archive::Iterator& rhs)
{
	return !operator==(lhs, rhs);
}

bool operator<(const Archive::Iterator& lhs, const Archive::Iterator& rhs)
{
	assert(lhs.archive_ != nullptr);
	assert(rhs.archive_ != nullptr);
	assert(lhs.archive_ == rhs.archive_);
	return lhs.file_index_ < rhs.file_index_;
}

bool operator<=(const Archive::Iterator& lhs, const Archive::Iterator& rhs)
{
	return operator<(lhs, rhs) || operator==(lhs, rhs);
}

bool operator>(const Archive::Iterator& lhs, const Archive::Iterator& rhs)
{
	return !operator<=(lhs, rhs);
}

bool operator>=(const Archive::Iterator& lhs, const Archive::Iterator& rhs)
{
	return operator>(lhs, rhs) || operator==(lhs, rhs);
}

}  // namespace sarc

File: cs/SimpleArchive/Archive.cs
//
// Copyright (c) 2016 wkmghn.
// 
// Use, modification and distribution is subject to the Boost Software License,
// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
//
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;
using System.Diagnostics;

namespace SimpleArchive
{
    /// <summary>
    /// Archive クラスの使用方法を表します。
    /// </summary>
    public enum ArchiveMode
    {
        /// <summary>
        /// アーカイブを新規作成します。
        /// Archive のコンストラクタに渡すストリームは空である必要があります。
        /// Archive がサポートするすべての機能を利用できます。
        /// </summary>
        Create,

        /// <summary>
        /// 既存のアーカイブを読み取り専用として扱います。
        /// Archive のコンストラクタに渡すストリームは既存のアーカイブを含んでいる必要があります。
        /// 書き込みを伴う一部の機能は利用できません。
        /// </summary>
        Read,

        /// <summary>
        /// 既存のアーカイブを変更します。
        /// Archive のコンストラクタに渡すストリームは既存のアーカイブを含んでいる必要があります。
        /// Archive がサポートするすべての機能を利用できます。
        /// </summary>
        Update,
    }

    /// <summary>
    /// アーカイブを表します。
    /// </summary>
    public class Archive : IDisposable
    {
        // 's', 'a', 'r', 'c'
        private readonly byte[] MAGIC_NUMBER = new byte[] { 0x73, 0x61, 0x72, 0x63 };
        private const UInt32 FILE_HEADER_SIZE = 4 + 4 + 4;  // MagicNumber + Version + NumFiles
        private const UInt32 MINIMUM_FILE_SIZE = FILE_HEADER_SIZE;
        private const UInt32 CURRENT_VERSION = 1;

        public IReadOnlyCollection<ArchiveEntry> Entries
        {
            get
            {
                if (m_isDisposed) { throw new ObjectDisposedException("Archive"); }
                return m_entries;
            }
        }

        /// <summary>
        /// アーカイブのモードを取得します。
        /// </summary>
        public ArchiveMode Mode { get { return m_mode; } }

        /// <summary>
        /// アーカイブ内のファイルにデフォルトで適用されるアライメントを取得します。
        /// </summary>
        public UInt32 DefaultAlignment { get { return m_defaultAlignment; } }

        /// <summary>
        /// 指定したストリームからアーカイブを初期化します。
        /// <para>
        /// ストリームの CanWrite が true の場合、ArchiveMode は Update に設定されます。
        /// それ以外の場合、ArchiveMode は Read に設定されます。
        /// </para>
        /// </summary>
        /// <param name="stream">
        /// 読み取るアーカイブを格納しているストリーム。
        /// ストリームは少なくともシークと読み取りをサポートしている必要があります。
        /// 指定されたストリームは Archive が破棄される際に閉じられます。
        /// </param>
        public Archive(Stream stream)
            : this(stream, stream.CanWrite ? ArchiveMode.Update : ArchiveMode.Read, false)
        {
        }

        /// <summary>
        /// 指定したストリームとモードを使用してアーカイブを初期化します。
        /// </summary>
        /// <param name="stream">
        /// 読み取るアーカイブを格納しているストリームもしくは新規作成されたアーカイブを出力するストリーム。
        /// ストリームは少なくともシークと読み取りをサポートしている必要があります。
        /// 指定されたストリームは Archive が破棄される際に閉じられます。
        /// </param>
        /// <param name="mode">
        /// アーカイブの使用方法を示す列挙値。
        /// </param>
        public Archive(Stream stream, ArchiveMode mode)
            : this(stream, mode, false)
        {
        }

        /// <summary>
        /// 指定したストリームとモードを使用してアーカイブを初期化します。
        /// </summary>
        /// <param name="stream">
        /// 読み取るアーカイブを格納しているストリームもしくは新規作成されたアーカイブを出力するストリーム。
        /// ストリームは少なくともシークと読み取りをサポートしている必要があります。
        /// </param>
        /// <param name="mode">
        /// アーカイブの使用方法を示す列挙値。
        /// </param>
        /// <param name="leaveOpen">
        /// アーカイブが破棄されてもストリームを開いたままにする場合は true を指定します。
        /// それ以外の場合は false を指定します。
        /// </param>
        /// <param name="dataAlignment">
        /// アーカイブ内のファイルにデフォルトで適用されるアライメントを取得します。
        /// この値はエントリごとにオーバーライドできます。
        /// 2 のべき乗数を指定する必要があります。
        /// </param>
        public Archive(Stream stream, ArchiveMode mode, bool leaveOpen, UInt32 defaultAlignment = 8)
        {
            if (stream == null)
            {
                throw new ArgumentNullException("stream");
            }
            if (!stream.CanRead)
            {
                throw new ArgumentException("読み込みをサポートするストリームが必要です。", "stream");
            }
            if (!stream.CanSeek)
            {
                throw new ArgumentException("シークをサポートするストリームが必要です。", "stream");
            }
            if (mode == ArchiveMode.Create || mode == ArchiveMode.Update)
            {
                if (!stream.CanWrite)
                {
                    throw new ArgumentException("書き込みをサポートするストリームが必要です。", "stream");
                }
            }
            if (!Misc.CheckExp2(defaultAlignment))
            {
                throw new ArgumentException("dataAlignment が二のべき乗数ではありません。", "dataAlignment");
            }

            m_stream = stream;
            m_mode = mode;
            m_closeStream = !leaveOpen;
            m_entries = new List<ArchiveEntry>();
            m_defaultAlignment = defaultAlignment;

            if (m_mode == ArchiveMode.Read || m_mode == ArchiveMode.Update)
            {
                // 既存のアーカイブを読み込む
                try
                {
                    if (m_stream.Length < MINIMUM_FILE_SIZE)
                    {
                        throw new CorruptedArchiveException("ストリームの長さが短すぎます。");
                    }

                    using (Reader reader = new Reader(m_stream, true))
                    {
                        // マジックナンバーをチェック
                        byte[] magicNumber = reader.ReadBytes(0, 4);
                        if (!magicNumber.SequenceEqual(MAGIC_NUMBER))
                        {
                            throw new CorruptedArchiveException("アーカイブのマジックナンバーが不正です。");
                        }

                        // バージョンをチェック
                        UInt32 version = reader.ReadUInt32(4);
                        if (version != CURRENT_VERSION)
                        {
                            throw new CorruptedArchiveException("アーカイブのバージョンが不正です。");
                        }

                        // アーカイブ内のファイル数を取得
                        UInt32 numFiles = reader.ReadUInt32(8);
                        // すべてのファイルを読み込む
                        for (UInt32 i = 0; i < numFiles; ++i)
                        {
                            ArchiveEntry entry = ReadEntry(reader, (int)i);
                            m_entries.Add(entry);
                        }
                    }
                }
                catch (Exception e)
                {
                    throw new CorruptedArchiveException("ストリームから読み込まれたアーカイブが壊れています。", e);
                }
            }
            else
            {
                // Mode == Create の場合。
                // ストリームが空っぽじゃないと困る。
                if (stream.Length != 0)
                {
                    throw new ArgumentException("ストリームが空ではありません。", "stream");
                }
            }
        }

        /// <summary>
        /// アーカイブ内の指定の名前のエントリを取得します。
        /// </summary>
        /// <param name="entryName">取得するエントリの名前。</param>
        /// <returns>アーカイブ内の指定された名前を持つエントリ。エントリがアーカイブ内に存在しない場合は null。</returns>
        /// <exception cref="ArgumentException">entryName が空文字列です。</exception>
        /// <exception cref="ArgumentNullException">entryName が null です。</exception>
        /// <exception cref="ObjectDisposedException">破棄された Archive にアクセスしようとしました。</exception>
        public ArchiveEntry GetEntry(string entryName)
        {
            if (m_isDisposed)
            {
                throw new ObjectDisposedException("Archive");
            }
            if (entryName == null)
            {
                throw new ArgumentNullException("entryName");
            }
            if (entryName == "")
            {
                throw new ArgumentException("entryName が空文字列です。", "entryName");
            }

            foreach (ArchiveEntry entry in m_entries)
            {
                if (entry.Name == entryName)
                {
                    return entry;
                }
            }

            return null;
        }

        /// <summary>
        /// 指定した名前を持つ空のエントリをアーカイブに作成します。
        /// </summary>
        /// <param name="entryName">作成されるエントリの名前。ASCII 文字だけで構成されている必要があります。</param>
        /// <returns>作成されたエントリ。</returns>
        /// <exception cref="ArgumentException">entryName が空文字列です。もしくは entryName に ASCII 以外の文字が含まれています。</exception>
        /// <exception cref="ArgumentNullException">entryName が null です。</exception>
        /// <exception cref="InvalidOperationException">
        /// 現在のアーカイブモードはエントリの作成をサポートしていません。
        /// もしくは
        /// 指定された名前と同名のエントリが既にアーカイブに存在します。
        /// </exception>
        /// <exception cref="ObjectDisposedException">破棄された Archive にアクセスしようとしました。</exception>
        public ArchiveEntry CreateEntry(string entryName)
        {
            if (m_isDisposed)
            {
                throw new ObjectDisposedException("Archive");
            }
            if (m_mode == ArchiveMode.Read)
            {
                throw new InvalidOperationException("現在のモードではエントリを作成できません。");
            }
            if (entryName == null)
            {
                throw new ArgumentNullException("entryName");
            }
            if (entryName == "")
            {
                throw new ArgumentException("entryName が空文字列です。", "entryName");
            }

            // entryName の文字コードをチェック
            try
            {
                Misc.ThrowIfContainsNoneASCIICharacters(entryName);
            }
            catch (EncoderFallbackException e)
            {
                throw new ArgumentException("entryName に ASCII 以外の文字が含まれています。", "entryName", e);
            }

            // 同名のエントリが既に無いかを確認
            foreach (var entry in m_entries)
            {
                if (entry.Name == entryName)
                {
                    throw new InvalidOperationException(string.Format("\"{0}\" という名前のエントリは既にアーカイブに存在します。", entryName));
                }
            }

            ArchiveEntry newEntry = new ArchiveEntry(this, entryName, null, DefaultAlignment, m_mode == ArchiveMode.Read);
            m_entries.Add(newEntry);
            return newEntry;
        }

        // entry をこのアーカイブの管理下から外す。
        // ArchiveEntry.Delete から呼ばれる想定。
        internal void RemoveEntryFromCollection(ArchiveEntry entry)
        {
            Debug.Assert(m_entries.Contains(entry));
            m_entries.Remove(entry);
        }

        // アーカイブの内容をストリームに書き出す。
        // m_mode が Read の場合は何もしない。
        private void Flush()
        {
            if (m_mode == ArchiveMode.Read)
            {
                return;
            }

            long archiveSize = 0;
            using (Writer writer = new Writer(m_stream, true))
            {
                writer.Seek(0, SeekOrigin.Begin);
                writer.WriteBytes(MAGIC_NUMBER);
                writer.WriteUInt32(CURRENT_VERSION);
                writer.WriteUInt32((UInt32)m_entries.Count);

                long offsetArraySize = 4 * m_entries.Count;

                // 一つ前のファイル末尾の、アーカイブ先頭からのオフセット
                // 現在のファイルはこのアドレスから書き込みを始める
                long prevFileEndAddress = FILE_HEADER_SIZE + offsetArraySize;
                for (int i = 0; i < m_entries.Count; ++i)
                {
                    ArchiveEntry entry = m_entries[i];
                    long fileHeaderAddress = prevFileEndAddress;

                    // ファイル先頭までオフセットを書き込む
                    {
                        long offsetAddress = FILE_HEADER_SIZE + 4 * i;
                        writer.Seek(offsetAddress, SeekOrigin.Begin);
                        writer.WriteUInt32((UInt32)fileHeaderAddress);
                    }

                    // ファイルヘッダとファイル本体を書き込む
                    {
                        writer.Seek(fileHeaderAddress, SeekOrigin.Begin);
                        // ヘッダの先頭はヘッダからファイル本体までのオフセットだが、一旦ダミー値を書き出す
                        writer.WriteUInt32(0xFFFFFFFF);
                        // ファイルサイズ
                        writer.WriteUInt32((UInt32)entry.Length);
                        writer.WriteUInt32(entry.Alignment);
                        // ファイル名
                        writer.WriteCStr(entry.Name);

                        // ファイル本体のアーカイブ内での位置をアラインする
                        long fileBodyAddressRaw = writer.Position;
                        long fileBodyAddressAligned = AlignAddress(fileBodyAddressRaw, entry.Alignment);
                        // ファイルヘッダ先頭に戻って、ヘッダからファイル本体へのオフセットを書き込む
                        // ファイルサイズがゼロでもそれらしい値が書き込まれる。
                        writer.Seek(fileHeaderAddress, SeekOrigin.Begin);
                        writer.WriteUInt32((UInt32)(fileBodyAddressAligned - fileHeaderAddress));

                        // ヘッダの末尾からアラインされたファイル本体までの隙間をゼロで埋める
                        // ファイルサイズがゼロでもパディングが書き出されるが、とりあえず気にしないでおく。
                        writer.Seek(fileBodyAddressRaw, SeekOrigin.Begin);
                        for (long address = fileBodyAddressRaw; address < fileBodyAddressAligned; ++address)
                        {
                            writer.WriteByte(0);
                        }
                        // 続いてファイル本体を書き出す
                        entry.WriteTo(writer);

                        // 次のファイルのアドレス計算のため、現在のファイルの末尾のアドレスを覚えておく
                        prevFileEndAddress = writer.Position;
                    }
                }

                archiveSize = prevFileEndAddress;
            }

            if (archiveSize < m_stream.Length)
            {
                // アーカイブが以前より小さくなったので、ストリームを切り詰める
                m_stream.SetLength(archiveSize);
            }
        }

        // stream から index 番目のエントリを読み取る。
        // stream の現在位置が変更される。
        private ArchiveEntry ReadEntry(Reader reader, int index)
        {
            Debug.Assert(reader != null);
            Debug.Assert(0 <= index);

            // オフセット配列へのオフセット
            UInt32 offsetOffset = FILE_HEADER_SIZE + ((UInt32)index * 4);
            // ファイルヘッダへのオフセット
            UInt32 fileHeadOffset = reader.ReadUInt32(offsetOffset);
            // ファイルの本体のバイト列へのオフセット
            UInt32 fileBodyOffset = fileHeadOffset + reader.ReadUInt32(fileHeadOffset);

            // ファイルサイズ
            UInt32 fileSize = reader.ReadUInt32(fileHeadOffset + 4);
            // アライメント
            UInt32 alignment = reader.ReadUInt32(fileHeadOffset + 8);
            // ファイル名
            string fileName = reader.ReadCStr(fileHeadOffset + 12);
            // ファイル本体
            byte[] data = reader.ReadBytes(fileBodyOffset, (int)fileSize);

            return new ArchiveEntry(this, fileName, data, alignment, m_mode == ArchiveMode.Read);
        }

        // address を二のべき乗数の倍数に切り上げる。
        // 切り上げ先のべき乗数は alignment で指定する。
        // alignment が二のべき乗数でない場合は例外を投げる。
        // また、address が負の数の場合も例外を投げる。
        private static long AlignAddress(long address, long alignment)
        {
            if (address < 0)
            {
                throw new ArgumentException("address が負の数です。", "address");
            }
            if (alignment < 0)
            {
                throw new ArgumentException("alignment が負の数です。", "alignment");
            }
            if (!Misc.CheckExp2(alignment))
            {
                throw new ArgumentException("alignment が二のべき乗数ではありません。", "alignment");
            }

            if ((address & (alignment - 1)) != 0)
            {
                // 切り上げて返す
                address = address & ~(alignment - 1);
                address = address + alignment;
                return address;
            }
            else
            {
                // address は alignment の倍数なので、そのまま返す
                return address;
            }
        }

        private Stream m_stream;
        private ArchiveMode m_mode;
        private bool m_closeStream;  // Dispose の際に m_stream を Close するか？
        private List<ArchiveEntry> m_entries;
        private UInt32 m_defaultAlignment;

#region IDisposable Support
        private bool m_isDisposed = false;

        protected virtual void Dispose(bool disposing)
        {
            if (!m_isDisposed)
            {
                Flush();
                if (disposing)
                {
                    if (m_stream != null && m_closeStream)
                    {
                        m_stream.Close();
                    }

                    foreach (ArchiveEntry entry in m_entries)
                    {
                        entry.NotifyArchiveDisposing();
                    }
                }

                m_stream = null;
                m_isDisposed = true;
            }
        }

        public void Dispose()
        {
            Dispose(true);
        }
#endregion
    }
}
